Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -1331,7 +1331,9 @@ protected class ClassWriter : IClassWriter {
     protected override ConcreteSyntaxTree EmitTailCallStructure(MemberDecl member, ConcreteSyntaxTree wr) {
       Contract.Assume((member is Method m0 && m0.IsTailRecursive) || (member is Function f0 && f0.IsTailRecursive)); // precondition
       if (!member.IsStatic && !NeedsCustomReceiver(member)) {
-        wr.WriteLine("var _this = this;");
+        var receiverType =  member.EnclosingClass is DatatypeDecl dt ?
+          "_I" + dt.CompileName + TypeParameters(SelectNonGhost(dt, dt.TypeArgs)) : "var";
```

Comment:
Oh I see, since the fields stores an interface, you need to provide the interface for the base type. Nice and easy fix.